src/vs/workbench/api/node/mainThreadLanguageFeatures.ts

```diff
@@ -137,13 +140,24 @@ export class MainThreadLanguageFeatures extends MainThreadLanguageFeaturesShape
 
 	$registerReferenceSupport(handle: number, selector: vscode.DocumentSelector): TPromise<any> {
 		this._registrations[handle] = modes.ReferenceProviderRegistry.register(selector, <modes.ReferenceProvider>{
-			provideReferences: (model: IReadOnlyModel, position: EditorPosition, context: modes.ReferenceContext, token: CancellationToken): Thenable<modes.Location[]> => {
-				return wireCancellationToken(token, this._proxy.$provideReferences(handle, model.uri, position, context));
+			provideReferences: (model: IReadOnlyModel, position: EditorPosition, context: modes.ReferenceContext, progress: (value: modes.Location) => void, token: CancellationToken): Thenable<modes.Location[]> => {
+				const progressHandle = this._progressRegistrations.registerChild(handle, progress);
+				let refs = this._proxy.$provideReferences(handle, progressHandle, model.uri, position, context);
+				refs = always(refs, () => this._progressRegistrations.unregisterChild(handle, progressHandle));
+				return wireCancellationToken(token, refs);
 			}
 		});
 		return undefined;
 	}
 
+	$provideReferenceProviderProgress(handle: number, progressHandle: number, location: modes.Location): TPromise<any> {
```

Comment:
Everything other method in this module is a `$register*` function, this doesn't follow the patterns. What purpose does this serve?